In [1]:
!pip install transformers

Hugging_face_Tasks

In [2]:
from transformers import pipeline
#----------------------------------------------------#
#                    NLP Tasks                       #
#----------------------------------------------------#
'''
1.Text Classification: Assinging a category to a piece of text.
Sentiment Analysis
Topic Classification
Spam Detection '''
classifier = pipeline("text-classification")

'''
2.Text Classification: Assinging  labels to individual tokens in a sequence.
Named Entity Recognition (NER)
Part-of-Speech Tagging  '''

token_classifier = pipeline("token-classification")

'''
3.Question Answering : Extracting as answer from a given context based on a question.
'''
question_answering = pipeline("question-answering")

'''
4.Text Generation : Generating text based on a given prompt.
Language Modeling
Story Generation

'''

text_generation = pipeline("text-generation")

'''
5.Summerization : Condensing long documents into shorter summaries
'''

summerizer = pipeline("summarization")

'''
6.Translation : Converting text from one language to another.
'''

translation = pipeline("translation",model="Helsinki-NLP/opus-mt-en-fr")

'''
7.Text2Text Generation : General-purpose text transformation , including summerization and translation.
'''

text2text_generation = pipeline("text2text-generation")

'''
8.Fill-Mask : predicting the masked token in a sequence.
'''

fill_mask = pipeline("fill-mask")

'''
9.Feature Extraction : Extracting hiddenn states or features from text.
'''
feature_extraction = pipeline("feature-extraction")


#----------------------------------------------------#
#               Computer Vision Tasks                #
#----------------------------------------------------#

'''
1.Image Classification : classifying the main content of an image.
'''
image_classifier = pipeline("image-classification")

'''
2.Object Detection : Identifying and localizing objects in an image.
'''
object_detector = pipeline("object-detection")

'''
3.Image Segmentation : Assigning a label to each pixel in an image.
'''
image_segmenter = pipeline("image-segmentation")


#----------------------------------------------------#
#              Speech Processing Tasks               #
#----------------------------------------------------#

'''
1.utomatic Speech Recognition (ASR) : Converting spoken language into text.
'''
speech_recognizer = pipeline("automatic-speech-recognition")

'''
2.Text-to-Speech (TTS) : Converting text into spoken language.
'''
text_to_speech = pipeline("text-to-speech")

'''
3.Speech Translation : Translating spoken language from one language to another.
4.Audio Classfication : Classifying audio signals into predefined categories.
'''

#----------------------------------------------------#
#                 Multimodal Tasks                   #
#----------------------------------------------------#

'''
1.Image Captioning: Generating a textual description of an image.
'''
image_captioner = pipeline("image-to-text")

'''
2.Visual Question Answering : Answering questions about the content of an image.
'''
visual_question_answerer = pipeline("visual-question-answering")
#----------------------------------------------------#
#                    other Tasks                     #
#----------------------------------------------------#
'''
1. Table Questioning Answering : Answering questions based on tabular data.
'''
table_qa=pipeline("table-question-answering")
'''
2.Document Question Answering : Extracting answers from documents like PDFs.
'''
doc_qa=pipeline("document-question-answering")
'''
3.Time Series Forecasting: Predicting future values in the series data .
'''

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision 4c53496 (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 564e9b5 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/301M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/42.0 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/301M [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/778k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/802k [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

/usr/local/lib/python3.11/dist-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")
Device set to use cuda:0
No model was supplied, defaulted to google-t5/t5-base and revision a9723ea (https://huggingface.co/google-t5/t5-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to distilbert/distilbert-base-cased and revision 6ea8117 (https://huggingface.co/distilbert/distilbert-base-cased).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/465 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/263M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to google/vit-base-patch16-224 and revision 3f49326 (https://huggingface.co/google/vit-base-patch16-224).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/346M [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/160 [00:00<?, ?B/s]

Fast image processor class <class 'transformers.models.vit.image_processing_vit_fast.ViTImageProcessorFast'> is available for this model. Using slow image processor class. To use the fast image processor class set `use_fast=True`.
Device set to use cuda:0
No model was supplied, defaulted to facebook/detr-resnet-50 and revision 1d5f47b (https://huggingface.co/facebook/detr-resnet-50).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/167M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/102M [00:00<?, ?B/s]

/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py:2397: UserWarning: for conv1.weight: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py:2397: UserWarning: for bn1.weight: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py:2397: UserWarning: for bn1.bias: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pas

preprocessor_config.json:   0%|          | 0.00/290 [00:00<?, ?B/s]

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.
Device set to use cuda:0
No model was supplied, defaulted to facebook/detr-resnet-50-panoptic and revision d53b52a (https://huggingface.co/facebook/detr-resnet-50-panoptic).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/172M [00:00<?, ?B/s]

Some weights of the model checkpoint at facebook/detr-resnet-50-panoptic were not used when initializing DetrForSegmentation: ['detr.model.backbone.conv_encoder.model.layer1.0.downsample.1.num_batches_tracked', 'detr.model.backbone.conv_encoder.model.layer2.0.downsample.1.num_batches_tracked', 'detr.model.backbone.conv_encoder.model.layer3.0.downsample.1.num_batches_tracked', 'detr.model.backbone.conv_encoder.model.layer4.0.downsample.1.num_batches_tracked']
- This IS expected if you are initializing DetrForSegmentation from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DetrForSegmentation from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


preprocessor_config.json:   0%|          | 0.00/289 [00:00<?, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to facebook/wav2vec2-base-960h and revision 22aad52 (https://huggingface.co/facebook/wav2vec2-base-960h).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/172M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/378M [00:00<?, ?B/s]

Some weights of Wav2Vec2ForCTC were not initialized from the model checkpoint at facebook/wav2vec2-base-960h and are newly initialized: ['wav2vec2.masked_spec_embed']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/163 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/291 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/85.0 [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/159 [00:00<?, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to suno/bark-small and revision 1dbd7a1 (https://huggingface.co/suno/bark-small).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.68G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.68G [00:00<?, ?B/s]

generation_config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/353 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to ydshieh/vit-gpt2-coco-en and revision 5bebf1e (https://huggingface.co/ydshieh/vit-gpt2-coco-en).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

/usr/local/lib/python3.11/dist-packages/transformers/models/auto/modeling_auto.py:2199: FutureWarning: The class `AutoModelForVision2Seq` is deprecated and will be removed in v5.0. Please use `AutoModelForImageTextToText` instead.
  warnings.warn(


pytorch_model.bin:   0%|          | 0.00/982M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/982M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/236 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/120 [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/211 [00:00<?, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to dandelin/vilt-b32-finetuned-vqa and revision d0a1f6a (https://huggingface.co/dandelin/vilt-b32-finetuned-vqa).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/470M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/320 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/470M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/251 [00:00<?, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to google/tapas-base-finetuned-wtq and revision e3dde19 (https://huggingface.co/google/tapas-base-finetuned-wtq).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

TAPAS models are not usable since `tensorflow_probability` can't be loaded. It seems you have `tensorflow_probability` installed with the wrong tensorflow version. Please try to reinstall it following the instructions here: https://github.com/tensorflow/probability.


pytorch_model.bin:   0%|          | 0.00/443M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/490 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/443M [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/154 [00:00<?, ?B/s]

Device set to use cuda:0
No model was supplied, defaulted to impira/layoutlm-document-qa and revision beed3c4 (https://huggingface.co/impira/layoutlm-document-qa).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/789 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/511M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/315 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

Device set to use cuda:0


'\n3.Time Series Forecasting: Predicting future values in the series data .\n'

In [3]:
'''
10.Text Similarity : Measuring the similarity between two texts.
'''
text_similarity = pipeline("text-similarity")

'''

4.Image Generation : Creating new images based on a given prompt.
'''
image_generator= pipeline("image-generation")

'''
5.Image Captioning : Generating a textual description of an image.
'''
image_captioning = pipeline("image-captioning")


KeyError: "Unknown task text-similarity, available tasks are ['audio-classification', 'automatic-speech-recognition', 'depth-estimation', 'document-question-answering', 'feature-extraction', 'fill-mask', 'image-classification', 'image-feature-extraction', 'image-segmentation', 'image-text-to-text', 'image-to-image', 'image-to-text', 'mask-generation', 'ner', 'object-detection', 'question-answering', 'sentiment-analysis', 'summarization', 'table-question-answering', 'text-classification', 'text-generation', 'text-to-audio', 'text-to-speech', 'text2text-generation', 'token-classification', 'translation', 'video-classification', 'visual-question-answering', 'vqa', 'zero-shot-audio-classification', 'zero-shot-classification', 'zero-shot-image-classification', 'zero-shot-object-detection', 'translation_XX_to_YY']"

NLP Tasks


1.SENTIMENT ANALYSIS

In [4]:
from transformers import pipeline
classifier =pipeline('sentiment-analysis')
res = classifier('I am happy to learn NLP')
print(res)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'POSITIVE', 'score': 0.9998495578765869}]


In [5]:
pipeline(task='sentiment-analysis')('I was confused with Barbie movie')

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'NEGATIVE', 'score': 0.9990049004554749}]

In [6]:
pipeline(task='sentiment-analysis')\
 ("Everyday lots of LLMs papers are published about LLMs Evaluation.\
 Lots of them looks very Promising.\
 I am not sure if we CAN actually EValuate LLMs.\
 There is still lot to do.\
 Don't you think? ")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'POSITIVE', 'score': 0.9784755706787109}]

In [7]:
#LM model for specific task
pipeline(task='sentiment-analysis',model='facebook/bart-large-mnli')\
 ("Everyday lots of LLMs papers are published about LLMs Evaluation.\
 Lots of them looks very Promising.\
 I am not sure if we CAN actually EValuate LLMs.\
 There is still lot to do.\
 Don't you think? ")

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


[{'label': 'neutral', 'score': 0.9840484857559204}]

Batch Sentiment Analysis

In [8]:
classifier = pipeline('sentiment-analysis')
res = classifier(["I am happy to learn NLP",
                  "I was confused with Barbie movie"])
print(res)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'POSITIVE', 'score': 0.9998495578765869}, {'label': 'NEGATIVE', 'score': 0.9990049004554749}]


In [9]:
#for getting multiple sentiments
classifier = pipeline('sentiment-analysis',model='SamLowe/roberta-base-go_emotions')
res = classifier(["I am happy to learn NLP",
                  "I was confused with Barbie movie"])
print(res)

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/380 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

Device set to use cuda:0


[{'label': 'joy', 'score': 0.8995077610015869}, {'label': 'confusion', 'score': 0.9352037906646729}]


Text Generation

In [10]:
#use a pipeline as a high-level helper
generator = pipeline(task='text-generation',model='distilbert/distilgpt2')
res = generator('I am happy to learn NLP',truncation=True,num_return_sequences= 2 )
print("Generated_text: \n",res[0]['generated_text'])

config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Generated_text: 
 I am happy to learn NLP has been an absolute success. I’m thrilled to work with the people behind NLP and have been very excited to share their vision of the future. The project has been a huge success, and I’m very proud of it.


I’m excited to help people with the creative spirit of this project.
I’m excited that NLP will allow them to be able to share their vision of the future.
I hope that NLP inspires people to do the best they can.
I’m excited to learn how to create and make a living by making something different.
I’m excited to learn how to create something different.
I’m happy to see NLP as a leading medium for learning about the future.
NLP is about creating great things.
The NLP platform is a great way to start new businesses.
I’m excited to learn how to create and make a living by making something different.
I’m happy to learn how to create and make a living by making something different.
I’m happy to learn how to create and make a living by making somethin

Question Answering

In [11]:
qa_model=pipeline("question-answering")
question="What is a decoder in transformer?"
context = "I am learning generating AI from scratch."
qa_model(question=question,context=context)

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 564e9b5 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


{'score': 0.9787534475326538, 'start': 25, 'end': 27, 'answer': 'AI'}

In [12]:
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification,DistilBertTokenizer,DistilBertForSequenceClassification

In [13]:
# First model: DistilBERT SST-2
model_name1 = 'distilbert-base-uncased-finetuned-sst-2-english'
tokenizer1 = AutoTokenizer.from_pretrained(model_name1)
model1 = AutoModelForSequenceClassification.from_pretrained(model_name1)

classifier1 = pipeline("sentiment-analysis", model=model1, tokenizer=tokenizer1)
res1 = classifier1([
    "I am happy to learn NLP",
    "I was confused with Barbie movie"
])
print("SST-2 Results:", res1)

# Second model: nlptown multilingual BERT
model_name2 = 'nlptown/bert-base-multilingual-uncased-sentiment'
tokenizer2 = AutoTokenizer.from_pretrained(model_name2)
model2 = AutoModelForSequenceClassification.from_pretrained(model_name2)

classifier2 = pipeline("sentiment-analysis", model=model2, tokenizer=tokenizer2)
res2 = classifier2([
    "I am happy to learn NLP",
    "I was confused with Barbie movie"
])
print("NLPTown Results:", res2)


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Device set to use cuda:0


SST-2 Results: [{'label': 'POSITIVE', 'score': 0.9998495578765869}, {'label': 'NEGATIVE', 'score': 0.9990049004554749}]


tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

Device set to use cuda:0


NLPTown Results: [{'label': '5 stars', 'score': 0.4859347939491272}, {'label': '3 stars', 'score': 0.2661842703819275}]


How tokenizer is working

In [14]:
from transformers import AutoTokenizer

#Load a pre trained tokenizer
tokenizer=AutoTokenizer.from_pretrained('bert-base-uncased')
#Text
text='I was so not happy with the Barbie movie'
#tokenize the text
tokens=tokenizer.tokenize(text)
print(tokens)
#Convert the tokens into input IDs or vector representation of my text
input_ids=tokenizer.convert_tokens_to_ids(tokens)
print("Representation of my text : ",input_ids)
#Encode the text (tokenization + converting the input IDs)
encoded_input=tokenizer(text)
print('Encoded Input : ',encoded_input)
#Decoded the text
decoded_output=tokenizer.decode(input_ids)
print('Decoded Output : ',decoded_output)
#

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

['i', 'was', 'so', 'not', 'happy', 'with', 'the', 'barbie', 'movie']
Representation of my text :  [1045, 2001, 2061, 2025, 3407, 2007, 1996, 22635, 3185]
Encoded Input :  {'input_ids': [101, 1045, 2001, 2061, 2025, 3407, 2007, 1996, 22635, 3185, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
Decoded Output :  i was so not happy with the barbie movie


1)101 : represents start of the sentence.(sos)


2)102 : represents end of the sentence.(eos)

3)attention_mask --> The attention mechanism concept that means whenever we pass the data to the model in which the model will give more weight

4)token_type_ids -->
These IDs are used to distinguish between different sequences in tasks that involve multiple sentences, such as question-answering and sentence-pair classification. BERT uses this mechanism to understand which tokens belong to which segment. For single-sequence tasks like sentiment analysis, token_type_ids are all zeros.

5)attention_mask-->
The attention mask is used to differentiate between actual tokens and padding tokens (if any). It helps the model focus on non-padding tokens and ignore padding tokens, A value of 1 indicates that the token should be attended to, while a value of 0 indicates padding.


6)Why Padding Tokens Are Used
Uniform Sequence Length--> Deep learning models typically process input data in batches. To efficiently process these batches, all sequences in a batch müst have the same length. Padding tokens ensure this by extending shorter sequences to match the length of the longest sequence in the batch. Efficient Computation: Fixed-length sequences allow for more efficient use of hardware resources, as the model can process all sequences in paralle without needing to handle variable-length sequences individually.

**Fine Tuning IMDB**

step1) install necessary libraries

In [15]:
!pip install --upgrade datasets transformers

step2) load the data

Upgrade to latest --> !pip install --upgrade datasets transformers

In [79]:
from datasets import load_dataset
dataset = load_dataset("imdb")


In [101]:
dataset

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 200
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 50
    })
    unsupervised: Dataset({
        features: ['text', 'label'],
        num_rows: 50000
    })
})

In [80]:
dataset["train"] = dataset["train"].shuffle(seed=42).select(range(200))
dataset["test"] = dataset["test"].shuffle(seed=42).select(range(50))

*Step 3 ) preprosessing the data*

In [86]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')

# Define tokenization function
def tokenize_function(examples):
    return tokenizer(
        examples['text'],
        padding='longest',   # pads all sequences to the same length
        truncation=True         # cuts off sequences longer than max_length
    )

# Apply tokenization to dataset
tokenized_datasets = dataset.map(tokenize_function, batched=True)


Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Map:   0%|          | 0/50 [00:00<?, ? examples/s]

Map:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [88]:
print(tokenized_datasets["train"][0])
print(tokenized_datasets["test"][0])


{'text': 'There is no relation at all between Fortier and Profiler but the fact that both are police series about violent crimes. Profiler looks crispy, Fortier looks classic. Profiler plots are quite simple. Fortier\'s plot are far more complicated... Fortier looks more like Prime Suspect, if we have to spot similarities... The main character is weak and weirdo, but have "clairvoyance". People like to compare, to judge, to evaluate. How about just enjoying? Funny thing too, people writing Fortier looks American but, on the other hand, arguing they prefer American series (!!!). Maybe it\'s the language, or the spirit, but I think this series is more English than American. By the way, the actors are really good and funny. The acting is not superficial at all...', 'label': 1, 'input_ids': [101, 1247, 1110, 1185, 6796, 1120, 1155, 1206, 3144, 2852, 1105, 26890, 1197, 1133, 1103, 1864, 1115, 1241, 1132, 2021, 1326, 1164, 5973, 6969, 119, 26890, 1197, 2736, 19501, 1183, 117, 3144, 2852, 273

In [102]:
tokenized_datasets["train"][0]

{'text': 'There is no relation at all between Fortier and Profiler but the fact that both are police series about violent crimes. Profiler looks crispy, Fortier looks classic. Profiler plots are quite simple. Fortier\'s plot are far more complicated... Fortier looks more like Prime Suspect, if we have to spot similarities... The main character is weak and weirdo, but have "clairvoyance". People like to compare, to judge, to evaluate. How about just enjoying? Funny thing too, people writing Fortier looks American but, on the other hand, arguing they prefer American series (!!!). Maybe it\'s the language, or the spirit, but I think this series is more English than American. By the way, the actors are really good and funny. The acting is not superficial at all...',
 'label': 1,
 'input_ids': [101,
  1247,
  1110,
  1185,
  6796,
  1120,
  1155,
  1206,
  3144,
  2852,
  1105,
  26890,
  1197,
  1133,
  1103,
  1864,
  1115,
  1241,
  1132,
  2021,
  1326,
  1164,
  5973,
  6969,
  119,
  

step 4 )setup training arguments

In [95]:
from transformers import TrainingArguments

training_args=TrainingArguments(output_dir='./results', #Output directory
                                eval_strategy="epoch",  #Evaluate every epoch
                                learning_rate=2e-5,     #Learning rate
                                per_device_train_batch_size=4, #Batch size for training
                                per_device_eval_batch_size=4, #Batch size for evaluation
                                num_train_epochs=3, #number of training epochs
                                weight_decay=0.01 # Strength of weight decay
)
training_args

TrainingArguments(
_n_gpu=1,
accelerator_config={'split_batches': False, 'dispatch_batches': None, 'even_batches': True, 'use_seedable_sampler': True, 'non_blocking': False, 'gradient_accumulation_kwargs': None, 'use_configured_state': False},
adafactor=False,
adam_beta1=0.9,
adam_beta2=0.999,
adam_epsilon=1e-08,
auto_find_batch_size=False,
average_tokens_across_devices=False,
batch_eval_metrics=False,
bf16=False,
bf16_full_eval=False,
data_seed=None,
dataloader_drop_last=False,
dataloader_num_workers=0,
dataloader_persistent_workers=False,
dataloader_pin_memory=True,
dataloader_prefetch_factor=None,
ddp_backend=None,
ddp_broadcast_buffers=None,
ddp_bucket_cap_mb=None,
ddp_find_unused_parameters=None,
ddp_timeout=1800,
debug=[],
deepspeed=None,
disable_tqdm=False,
do_eval=True,
do_predict=False,
do_train=False,
eval_accumulation_steps=None,
eval_delay=0,
eval_do_concat_batches=True,
eval_on_start=False,
eval_steps=None,
eval_strategy=IntervalStrategy.EPOCH,
eval_use_gather_object=False

Step 5 ) Initialize the model

In [96]:
import torch, gc
gc.collect()
torch.cuda.empty_cache()


In [103]:
from transformers import AutoModelForSequenceClassification, Trainer

# Load the pre-trained model
model = AutoModelForSequenceClassification.from_pretrained('distilbert-base-uncased', num_labels=2)
#Initialize the trainer
trainer=Trainer(model=model,args=training_args,train_dataset=tokenized_datasets['train'],eval_dataset=tokenized_datasets['test'])

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Step 6 ) Train the model

In [104]:
#Train the model
trainer.train()

TrainOutput(global_step=150, training_loss=0.6942327372233073, metrics={'train_runtime': 48.8651, 'train_samples_per_second': 12.279, 'train_steps_per_second': 3.07, 'total_flos': 79480439193600.0, 'train_loss': 0.6942327372233073, 'epoch': 3.0})

Step 7 ) Evaluate the model

In [105]:
#Evaluate the model
results = trainer.evaluate()
print(results)

{'eval_loss': 0.6900845170021057, 'eval_runtime': 0.822, 'eval_samples_per_second': 60.828, 'eval_steps_per_second': 15.815, 'epoch': 3.0}


Step 8 ) Save the fine tuned model

In [106]:
#save the model
model.save_pretrained('./fine_tuned-model')
tokenizer.save_pretrained('./fine_tuned-model')

('./fine_tuned-model/tokenizer_config.json',
 './fine_tuned-model/special_tokens_map.json',
 './fine_tuned-model/vocab.txt',
 './fine_tuned-model/added_tokens.json',
 './fine_tuned-model/tokenizer.json')